### PR TITLE
feat: condition Audit log on licensing

### LINKED ISSUE #3219: Show Audit Log if licensed
## Problem
As an enterprise customer, I want to see a UI of the audit log if I'm licensed for it.

## Definition of Done

- [ ] have the audit log route go to the Audit Log Page if the license allows it, else to the not found page
- [ ] make the nav link show only if the license allows it

## Notes
See #3008 for implementation ideas, such as a wrapper component for licensed pages and a selector in the `licenseXService`

### PR DESCRIPTION
**Pull request by presleyp**
Closes #3219 

- Add `RequirePermissions` wrapper component to make `AppRouter` more readable (currently doing the XState stuff in `AppRouter` instead of the wrapper, either way is workable though)
- Condition Audit Log link and page visibility on license in addition to RBAC and dev mode
- upgrade xstate-react because a recent patch fixes a testing issue

In the past, we've punted on some component tests thinking that e2e tests would work better for cases that need to include the whole component tree. But e2e tests are difficult - they require setup instead of mocks and can be flaky. So I tried doing a component test while rendering `App`. I think it's a good way forward for such cases.

Click SHOW_MOCK_BANNER and HIDE_MOCK_BANNER in the XState inspector to see/hide the Audit link in the navbar - they change the stored license data to trigger both the warning banner and enabling the Audit Log.

Remaining issue:
A test that passes in my workspace fails on CI, I'm investigating.

### CODE REVIEW

File: site/package.json [hunk 1]
```diff
@@ -33,7 +33,7 @@
     "@material-ui/lab": "4.0.0-alpha.42",
     "@testing-library/react-hooks": "8.0.1",
     "@xstate/inspect": "0.6.5",
-    "@xstate/react": "3.0.0",
+    "@xstate/react": "3.0.1",
```
Comment (jsjoeio) [inline site/package.json:36]: What was the testing issue this fixed?

File: site/src/api/types.ts [hunk 2]
```diff
@@ -14,3 +14,9 @@ export interface ReconnectingPTYRequest {
 export type WorkspaceBuildTransition = "start" | "stop" | "delete"
 
 export type Message = { message: string }
+
+// Keep up to date with coder/codersdk/features.go
```
Comment (jsjoeio) [inline site/src/api/types.ts:18]: Is this something that can be autogenerated?

File: site/src/components/Navbar/Navbar.test.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+describe("Navbar", () => {
+  it("shows Audit Log link when permitted and entitled", async () => {
+    server.use(
+      rest.get("/api/v2/entitlements", (req, res, ctx) => {
+        return res(ctx.status(200), ctx.json(MockEntitlementsWithAuditLog))
+      }),
+    )
+    render(<App />)
+    await waitFor(() => {
+      const link = screen.getByText(Language.users) // TODO change after debugging
```
Comment (jsjoeio) [inline site/src/components/Navbar/Navbar.test.tsx:25]: Did you figure this out?

File: site/src/components/Navbar/Navbar.test.tsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      expect(link).toBeDefined()
+    })
+  })
+
+  it("does not show Audit Log link when not entitled", async () => {
```
Comment (jsjoeio) [inline site/src/components/Navbar/Navbar.test.tsx:30]: Reading this code, how does one know Audit Log link is not entitled? Is it due to the lack of the mocked request or something?

File: site/src/components/Navbar/Navbar.test.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    render(<App />)
+    await waitFor(() => {
+      const link = screen.queryByText(Language.audit)
+      expect(link).toBe(null)
+    })
+  })
+
+  it("does not show Audit Log link when not permitted via role", async () => {
```
Comment (jsjoeio) [inline site/src/components/Navbar/Navbar.test.tsx:38]: Same with this test. How does one know that a member is not permitted? I see `MockMemberPermissions` but hard to tell from code what role is being given or why they don't have permission.

File: site/src/components/RequirePermission/RequirePermission.tsx [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import { FC } from "react"
+import { Navigate } from "react-router"
+
+export interface RequirePermissionProps {
+  children: JSX.Element
```
Comment (jsjoeio) [inline site/src/components/RequirePermission/RequirePermission.tsx:5]: Would [`PropsWithChildren`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L799) help here? [reference](https://www.mendelowski.com/docs/react/children-props-with-typescript/)

File: site/src/testHelpers/entities.ts [hunk 7]
```diff
@@ -51,6 +51,10 @@ export function assignableRole(role: TypesGen.Role, assignable: boolean): TypesG
   }
 }
 
+export const MockMemberPermissions = {
```
Comment (jsjoeio) [inline site/src/testHelpers/entities.ts:54]: Might be helpful to convey permissions in name i.e. `MockMemberWithoutAuditLogPermissions`

File: site/package.json (hunk 1 above)
Comment (presleyp) [reply site/package.json:36]: The `getSnapshot` warning/crash and the `shallowEqual` util for avoiding it. https://newreleases.io/project/github/statelyai/xstate/release/@xstate%2Freact@3.0.1

File: site/src/api/types.ts (hunk 2 above)
Comment (presleyp) [reply site/src/api/types.ts:18]: My understanding is there's some reason why we can't get enums in our generated types.

File: site/src/components/Navbar/Navbar.test.tsx (hunk 4 above)
Comment (presleyp) [reply site/src/components/Navbar/Navbar.test.tsx:30]: Yes, the default handler in `handlers.ts` returns empty license data, as if you're an OSS user.

File: site/src/components/Navbar/Navbar.test.tsx (hunk 5 above)
Comment (presleyp) [reply site/src/components/Navbar/Navbar.test.tsx:38]: The `Member` role does not have auditing permissions.

File: site/src/testHelpers/entities.ts (hunk 7 above)
Comment (presleyp) [reply site/src/testHelpers/entities.ts:54]: I added comments, does that help?

Comment (jsjoeio) [reply site/src/testHelpers/entities.ts:54]: Yes, it does! Thank you!

**Review by jsjoeio — APPROVED** [review]
Nice work!

File: site/src/components/RequirePermission/RequirePermission.tsx (hunk 6 above)
Comment (presleyp) [reply site/src/components/RequirePermission/RequirePermission.tsx:5]: I tried using it here and got a TS error, not sure why!

Comment (jsjoeio) [reply site/src/components/RequirePermission/RequirePermission.tsx:5]: This works in the playground: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDGMA0cDecAKUEYAzgOrAwAWAwlcADYAmUSAdnAL5wBmRIcAOSt0MQQFgAUFKQAPSLDgwAnmCSIkARwCuwVviRQQwEiWAQ2hYiTgBeAkVIVqdRi3YAeHFLhwTAMVQYbVYANRNgACMGJAAuOEiICBiUNilOAD4pKTQLEngAORQAN2AAcxQYdXsACjBHEnicGAh4-KhgNjLOAEo7DLgPFDgqVh5bHHrrADoWzIZOgGsPAHoULOlJFYAqbZ9tuDIoFFI4Im0qm2pKuBRWW+KURhRo9UiUEiQmOAtEACEAII0H5QOALNDsMxdab7FYyeTQeC5Nj5DQ6PRIAxGExmX61PBoejMVhsbABIIhJDhMyvLjxZDo-SGYymcyWBp9WwDbySXzAHhwGoAQnJlUp1KiMT6PN8vlYwSgHA8RVKFSqSggtgARCsAO7QRYkMDoJAkLVwFYbXzcJAMT64Hyy+UhDiEtwkx2cdJAA

Up to you though: just a suggestion

**Pull request merged**